ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs

```diff
@@ -0,0 +1,744 @@
+deriving instance ToExpr BlockNo
+deriving instance ToExpr SlotNo
+deriving instance ToExpr ReaderNext
+deriving instance ToExpr IteratorId
+deriving instance ToExpr (HeaderHash blk) => ToExpr (ChainHash blk)
+deriving instance ToExpr (HeaderHash blk) => ToExpr (Point blk)
+deriving instance ToExpr (HeaderHash blk) => ToExpr (ReaderState blk)
+deriving instance ToExpr blk => ToExpr (Chain blk)
+deriving instance ( ToExpr blk
+                  , ToExpr (HeaderHash blk)
+                  )
+                 => ToExpr (ChainProducerState blk)
+deriving instance ( ToExpr blk
+                  , ToExpr (HeaderHash blk)
+                  , ToExpr (ChainState (BlockProtocol blk))
+                  , ToExpr (LedgerState blk)
+                  )
+                 => ToExpr (DBModel blk)
+deriving instance ( ToExpr blk
+                  , ToExpr (HeaderHash blk)
+                  , ToExpr (ChainState (BlockProtocol blk))
+                  , ToExpr (LedgerState blk)
+                  )
+                 => ToExpr (Model blk hdr IO Concrete)
+
+-- Blk and Hdr specific instances
+
+deriving instance ToExpr TestHash
+deriving instance ToExpr Blk
+deriving instance ToExpr (LedgerState Blk)
+
+{-------------------------------------------------------------------------------
+  Top-level tests
+-------------------------------------------------------------------------------}
+
+type Blk = TestBlock
+type Hdr = TestBlock
+
+prop_sequential :: Property
+prop_sequential = forAllCommands smUnused Nothing $ \cmds -> QC.monadicIO $ do
+    (hist, prop) <- test cmds
+    -- TODO tags
+    prettyCommands smUnused hist prop
+  where
+    test :: QSM.Commands (At Cmd Blk Hdr IO) (At Resp Blk Hdr IO)
+         -> QC.PropertyM IO
+            ( QSM.History (At Cmd Blk Hdr IO) (At Resp Blk Hdr IO)
+            , Property
+            )
+    test cmds = do
+      args <- QC.run $ mkArgs cfg initLedger
+      db   <- QC.run $ openDB args
+      let sm' = sm db genBlock env initLedger
+      (hist, model, res) <- runCommands sm' cmds
+      realChain <- QC.run $ chainHashes <$> toChain db
+      QC.run $ closeDB db
+      let modelChain = chainHashes $ Model.currentChain $ dbModel model
+          prop =
+            counterexample ("Real  chain: " <> show realChain)  $
+            counterexample ("Model chain: " <> show modelChain) $
+            res === Ok .&&. realChain === modelChain
+      return (hist, prop)
+
+    cfg :: NodeConfig (BlockProtocol Blk)
+    cfg = singleNodeTestConfig
+
+    initLedger :: ExtLedgerState Blk
+    initLedger = testInitExtLedger
+
+    env :: ModelEnv Blk Hdr
+    env = ModelEnv { cfg = cfg, blockHeader = id }
+
+    smUnused = sm dbUnused genBlock env initLedger
+
+    -- TODO
+    genBlock :: Gen Blk
+    genBlock = do
+      fork <- choose (1, 4)
+      no   <- choose (1, 20)
+      let prevHash
+            | no == 1   = GenesisHash
+            | otherwise = BlockHash (TestHash (fork * 1000 + no - 1))
```

Comment:
Represent a path through a tree as a list of `Word64`s: `[0, 1, 0]` means we forked off at the 1st block.
```
G--A--B--C
\  \--B"-C"
 \-A'-B'-C'

[]      = G = Genesis
[0]     = A
[1]     = A'
[0,0]   = B
[0,0,0] = C
[0,1]   = B"
[0,1,0] = C"
[1,0]   = B'
[1,0,0] = C'
```

Reply:
The predecessor (parent in the tree) can be obtained by dropping the last element in the list.